File: app/src/main/java/org/mozilla/focus/home/PrivateHomeScreenBackground.java
/* -*- Mode: Java; c-basic-offset: 4; tab-width: 4; indent-tabs-mode: nil; -*-
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

package org.mozilla.focus.home;

import android.app.Activity;
import android.content.Context;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.graphics.BitmapShader;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.ComposeShader;
import android.graphics.LinearGradient;
import android.graphics.Paint;
import android.graphics.PorterDuff;
import android.graphics.Rect;
import android.graphics.Shader;
import android.support.annotation.Nullable;
import android.support.v7.widget.AppCompatImageView;
import android.util.AttributeSet;

import org.mozilla.focus.R;

public class PrivateHomeScreenBackground extends AppCompatImageView {
    private Paint paint;

    public PrivateHomeScreenBackground(Context context) {
        super(context, null);
        init();
    }

    public PrivateHomeScreenBackground(Context context, AttributeSet attrs) {
        super(context, attrs, 0);
        init();
    }

    public PrivateHomeScreenBackground(Context context, @Nullable AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        init();
    }

    void init() {
        Rect rect = new Rect();
        ((Activity) getContext()).getWindow().getDecorView().getWindowVisibleDisplayFrame(rect);
        Bitmap bitmap = BitmapFactory.decodeResource(getResources(), R.drawable.home_pattern);
        paint = new Paint();
        Shader shader1 = new BitmapShader(bitmap, Shader.TileMode.REPEAT, Shader.TileMode.REPEAT);
        int colors[] = {Color.parseColor("#99FFFFFF"), Color.parseColor("#4dFFFFFF"), Color.parseColor("#1aFFFFFF"), Color.parseColor("#00FFFFFF")};
        float positions[] = {0.0f, 0.4f, 0.7f, 1f};
        Shader shader2 = new LinearGradient(0, rect.top, 0, rect.bottom, colors, positions, Shader.TileMode.CLAMP);
        paint.setShader(new ComposeShader(shader2, shader1, PorterDuff.Mode.MULTIPLY));
    }

    @Override
    protected void onDraw(Canvas canvas) {
        super.onDraw(canvas);
        canvas.drawRect(0, 0, getWidth(), getHeight(), paint);
    }
}

File: app/src/main/java/org/mozilla/rocket/home/ui/PrivateHomeScreenBackground.kt
/* -*- Mode: Java; c-basic-offset: 4; tab-width: 4; indent-tabs-mode: nil; -*-
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

package org.mozilla.rocket.home.ui

import android.content.Context
import android.graphics.Bitmap
import android.graphics.Canvas
import android.graphics.LinearGradient
import android.graphics.Paint
import android.graphics.PorterDuff
import android.graphics.PorterDuffXfermode
import android.graphics.Shader
import android.graphics.drawable.Drawable
import androidx.appcompat.widget.AppCompatImageView
import android.util.AttributeSet
import androidx.appcompat.content.res.AppCompatResources
import androidx.core.content.ContextCompat
import androidx.core.graphics.drawable.toBitmap
import androidx.core.graphics.drawable.toDrawable

import org.mozilla.focus.R

class PrivateHomeScreenBackground : AppCompatImageView {
    constructor(context: Context) : super(context, null) {
        init(context)
    }

    constructor(context: Context, attrs: AttributeSet?) : super(context, attrs, 0) {
        init(context)
    }

    constructor(context: Context, attrs: AttributeSet?, defStyleAttr: Int) : super(context, attrs, defStyleAttr) {
        init(context)
    }

    internal fun init(context: Context) {
        scaleType = ScaleType.CENTER_CROP
        applyPrivateTheme(context)
    }

    override fun onDraw(canvas: Canvas) {
        super.onDraw(canvas)
        setPadding(0, 0, 0, 0)
    }

    private fun applyPrivateTheme(context: Context) {
        val colorStart = ContextCompat.getColor(context, R.color.themeGradientStartPrivate)
        val colorCenter = ContextCompat.getColor(context, R.color.themeGradientCenterPrivate)
        val colorEnd = ContextCompat.getColor(context, R.color.themeGradientEndPrivate)

        setImageDrawable(getPatternDrawable(colorStart, colorCenter, colorEnd))
    }

    private fun getPatternDrawable(colorStart: Int, colorCenter: Int, colorEnd: Int): Drawable =
            getScreenedDrawable(
                    PATTERN_IMAGE_RESOURCE_ID,
                    intArrayOf(colorStart, colorCenter, colorEnd),
                    COLORS_GRADIENT_POSITIONS
            )

    private fun getScreenedDrawable(resId: Int, colors: IntArray, positions: FloatArray): Drawable =
            applyColorScreen(
                    bitmap = requireNotNull(AppCompatResources.getDrawable(context, resId)).toBitmap(),
                    colors = colors,
                    positions = positions
            ).toDrawable(resources)

    private fun applyColorScreen(bitmap: Bitmap, colors: IntArray, positions: FloatArray): Bitmap {
        val width = bitmap.width
        val height = bitmap.height
        val updatedBitmap = Bitmap.createBitmap(width, height, Bitmap.Config.ARGB_8888)
        val canvas = Canvas(updatedBitmap)

        canvas.drawBitmap(bitmap, 0f, 0f, null)

        val paint = Paint()
        val shader = LinearGradient(0f, 0f, width.toFloat(), height.toFloat(), colors, positions, Shader.TileMode.CLAMP)
        paint.shader = shader
        paint.xfermode = PorterDuffXfermode(PorterDuff.Mode.SCREEN)
        canvas.drawRect(0f, 0f, width.toFloat(), height.toFloat(), paint)

        return updatedBitmap
    }

    companion object {
        private const val PATTERN_IMAGE_RESOURCE_ID = R.drawable.bg_homescreen_pattern
        private val COLORS_GRADIENT_POSITIONS = floatArrayOf(0f, 0.65f, 1f)
    }
}
